File: Assets/src/model/enemies/Crab.cs
using System;
using System.Collections.Generic;
using UnityEngine;

[System.Serializable]
[ObjectInfo(description: "Only moves horizontally.\nAttacks anything in its path.", flavorText: "")]
public class Crab : AIActor {
  public override float turnPriority => 40;
  public int dx { get; private set; }
  [field:NonSerialized] /// controller only
  public event Action OnDirectionChanged;

  public Crab(Vector2Int pos) : base(pos) {
    dx = MyRandom.value < 0.5 ? -1 : 1;
    faction = Faction.Neutral;
    hp = baseMaxHp = 7;
  }

  internal override (int, int) BaseAttackDamage() => (2, 2);

  protected override ActorTask GetNextTask() {
    var player = GameModel.main.player;

    var nextPos = pos + new Vector2Int(dx, 0);
    var nextTile = floor.tiles[nextPos];
    if (nextTile.BasePathfindingWeight() == 0 || nextTile.actor is Crab) {
      // can't walk there; change directions
      dx *= -1;
      OnDirectionChanged?.Invoke();
      return new WaitTask(this, 1);
    } else {
      if (nextTile.body == null) {
        return new MoveToTargetTask(this, nextPos);
      } else {
        // something's blocking the way; attack it
        return new AttackTask(this, nextTile.body);
      }
    }
  }
}

File: web/src/model/enemies/Crab.ts
import { AIActor } from './AIActor';
import { ActorTask } from '../ActorTask';
import { AttackTask } from '../tasks/AttackTask';
import { MoveToTargetTask } from '../tasks/MoveToTargetTask';
import { WaitTask } from '../tasks/WaitTask';
import { Vector2Int } from '../../core/Vector2Int';
import { Faction } from '../../core/types';
import { MyRandom } from '../../core/MyRandom';
import { entityRegistry } from '../../generator/entityRegistry';
import type { Body } from '../Body';

/**
 * Neutral creature. Only moves horizontally. Attacks anything in its path.
 * Port of C# Crab.cs.
 */
export class Crab extends AIActor {
  get turnPriority(): number {
    return 40;
  }

  dx: number;
  onDirectionChanged: (() => void) | null = null;

  constructor(pos: Vector2Int) {
    super(pos);
    this.dx = MyRandom.value < 0.5 ? -1 : 1;
    this.faction = Faction.Neutral;
    this.hp = this._baseMaxHp = 5;
  }

  baseAttackDamage(): [number, number] {
    return [2, 2];
  }

  protected getNextTask(): ActorTask {
    const nextPos = Vector2Int.add(this.pos, new Vector2Int(this.dx, 0));
    const nextTile = this.floor?.tiles.get(nextPos);

    if (!nextTile || nextTile.basePathfindingWeight() === 0 || (nextTile.body instanceof Crab)) {
      // Can't walk there; change directions
      this.dx *= -1;
      this.onDirectionChanged?.();
      return new WaitTask(this, 1);
    } else {
      if (nextTile.body == null) {
        return new MoveToTargetTask(this, nextPos);
      } else {
        // Something's blocking the way; attack it
        return new AttackTask(this, nextTile.body as Body);
      }
    }
  }
}

entityRegistry.register('Crab', Crab);
